src/ast/passes/semantic_analyser.cpp

```diff
@@ -2563,30 +2570,41 @@ void SemanticAnalyser::visit(For &f)
   //       $len++;
   //     }
 
-  // Validate decl
+  // Validate decl.
   const auto &decl_name = f.decl->ident;
   if (find_variable(decl_name)) {
     f.decl->addError() << "Loop declaration shadows existing variable: " +
                               decl_name;
   }
 
-  if (!f.map->type().IsMapIterableTy()) {
-    f.map->addError() << "Loop expression does not support type: "
-                      << f.map->type();
-    return;
+  visit(f.iterable);
+
+  // Validate the iterable.
+  if (auto *map = f.iterable.as<Map>()) {
+    if (!map->type().IsMapIterableTy()) {
+      map->addError() << "Loop expression does not support type: "
+                      << map->type();
+      return;
+    }
+  } else if (auto *range = f.iterable.as<Range>()) {
+    if (is_final_pass()) {
+      if (!range->start.type().IsIntTy()) {
+        range->addError() << "Loop range requires integer start";
```

Comment:
nit
```suggestion
        range->addError() << "Loop range requires an integer for the start";
```
(same for the end below)